### PR TITLE
Mint bird from LIFF

### PR DESCRIPTION
**Pull request by shiki-tak**
I implemented the following to make it possible to mint Birds as NFTs from the Flutter Bird App.

**flutter_bird_image_server**
- Implement flutter_bird_image_server
  - A bird image is generated based on the received tokenID.
  - The same data url will always be returned for the same tokenId.
  - Currently, the hosting is on Vercel, and there are restrictions on using the file system. Therefore, the generated image is not saved on the server, but the data URL is included in the response and returned.
  - The image size of bird is around 10KB at most, so it does not put a large load on the system.
- flutter_bird_image_server/create_birds/create_data.js and the generate_image and generate_metadata files it calls use the files that were already in the original repository. Only the parts related to fs have been deleted.
   - https://github.com/Tonnanto/flutter-bird/tree/master/flutter_bird_skins/scripts
- mint_full_onchain_bird/mint_random_skin.js is a script for manually minting bird and will not be called when minting from the flutter app.

 **flutter_bird_app**
- Implement mint button on flutter app
  - I added a button to mint bird on flutter.
  - This receives the data url from the image_server and calls mint and appendUri of the FlutterBirdsSkins contract.
- _createAndUploadUri
  - https://github.com/Finschia/flutter-bird/compare/master...Finschia:flutter-bird:mint-bird-from-liff?expand=1#diff-21f69922bf92a0ace86a77db5c1db6d4c9f67f3ea7adf9b3a8ca96e96ab264feR116
  - The image is retrieved from the server using the tokenId as a key, and the contract's appendUri is executed.
  - Currently, the game operator executes appendUri using a private key. This specification is merely a PoC to verify whether tx can be executed in a format different from mint, and is not meaningful.
- mintRandomSkin
  - https://github.com/Finschia/flutter-bird/compare/master...Finschia:flutter-bird:mint-bird-from-liff?expand=1#diff-21f69922bf92a0ace86a77db5c1db6d4c9f67f3ea7adf9b3a8ca96e96ab264feR35
  - Users will use 1 KLAY to mint an NFT.
  
**flutter_bird_skins**
- Implement appendUri
  - The image data url is set to the tokenURI of the NFT. 
  - This allows you to retrieve the image data by simply querying the contract's tokenURI. 
  - The image is not stored anywhere on the file server and exists entirely on-chain.
  - Therefore, we **no longer use IPFS**.
- lib contracts
  - The contracts in flutter_bird_skins/lib use already implemented contracts around storage.

**LIFF URL**
- You can try out mint with the LIFF below.
  - https://liff.line.me/2005806035-kp4pmzW2

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -16,7 +16,7 @@ RUN apt-get update && apt-get install -y \
 RUN npm install -g node-gyp
 RUN npm install -g truffle
 
-COPY ./flutter_bird_skins/package*.json ./
+COPY ./flutter_bird_skins/package.json ./
```
Comment (Mdaiki0730) [inline Dockerfile]: Do you not include the package-lock?

File: flutter_bird_image_server/create_birds/generate_metadata.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    const eyesList = traits.eyes;
+    const mouthList = traits.mouth;
+    const neckList = traits.neck;
+
+    let guaranteedTrait = Math.floor(rng() * 5) - 1;
+    const randomBird = getRandomWeightedTrait(birdList, false, rng);
+    if (randomBird === 'default') guaranteedTrait = Math.floor(rng() * 4);
+
+    let guaranteedNotTrait = Math.floor(rng() * 6);
+    if (guaranteedTrait === guaranteedNotTrait) guaranteedNotTrait = -1;
```
Comment (Mdaiki0730) [inline flutter_bird_image_server/create_birds/generate_metadata.js:48]: What do guaranteedTrait and guaranteedNotTrait refer to?

File: flutter_bird_image_server/server.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    console.error('Error in /api/image/:tokenId route:', e);
+    res.status(500).json({ error: 'Internal server error' });
+  }
+});
+
+app.get('/', (req, res) => {
+  res.status(200).json({ status: 'OK', message: 'Server is running' });
+});
+
+if (process.env.NODE_ENV !== 'production') {
```
Comment (Mdaiki0730) [inline flutter_bird_image_server/server.js:28]: What about for production?

File: flutter_bird_skins/contracts/FlutterBirdSkins.sol [hunk 4]
```diff
@@ -59,38 +62,23 @@ contract FlutterBirdSkins is ERC721Enumerable {
...
-    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
-        require(_exists(tokenId), "ERC721URIStorage: URI query for nonexistent token");
-
-        string memory _tokenURI = _tokenURIs[tokenId];
-        string memory base = _baseURI();
-
-        // If there is no base URI, return the token URI.
-        if (bytes(base).length == 0) {
-            return _tokenURI;
+    function appendUri(uint256 tokenId, bytes[] calldata values) public onlyOwner {
```
Comment (Mdaiki0730) [inline flutter_bird_skins/contracts/FlutterBirdSkins.sol:65]: Can there be multiple URIs for one token?

**Comment by Mdaiki0730** [comment]
I can't connect to metamask, why does it take me to the app store with the app downloaded if it is a universal link?

https://github.com/user-attachments/assets/b1fc8d52-4572-4bda-b662-9e8b7f3acf54

**Comment by shiki-tak** [comment]
> I can't connect to metamask, why does it take me to the app store with the app downloaded if it is a universal link?
> 
>  RPReplay_Final1721871875.mp4

I don't have iOS so I can't test it thoroughly. I look into it.

**Review by da1suk8 — DISMISSED** [review]
LGTM

**Comment by 170210** [comment]
I want ask a question. 
When I tried https://liff.line.me/2005806035-kp4pmzW2 with metamask, it requires me signature two times. But it doesn't appear in current LIFF application.

**Comment by shiki-tak** [comment]
> I want ask a question. When I tried https://liff.line.me/2005806035-kp4pmzW2 with metamask, it requires me signature two times. But it doesn't appear in current LIFF application.

When do you mean "two signatures"? When I tried it, it appeared once each on connect and mint. In the case of mint, if you manually navigate to metamask, the signature popup will appear.

File: flutter_bird_skins/contracts/FlutterBirdSkins.sol (hunk 4 above)
Comment (shiki-tak) [reply flutter_bird_skins/contracts/FlutterBirdSkins.sol:65]: No, one token is assigned to one URI.

File: Dockerfile (hunk 1 above)
Comment (shiki-tak) [reply Dockerfile]: Basically, the lock file is an automatically generated file, so I decided there is no need to include it here.

File: flutter_bird_image_server/create_birds/generate_metadata.js (hunk 2 above)
Comment (shiki-tak) [reply flutter_bird_image_server/create_birds/generate_metadata.js:48]: This variable determines how randomly each element of the randomly generated bird (head, mouth, etc.) will be generated.

guaranteedTrait is a variable that guarantees that a specific element will always be selected other than the default.
guaranteedNotTrait is a variable that guarantees that a specific element will be the default.

File: flutter_bird_image_server/server.js (hunk 3 above)
Comment (shiki-tak) [reply flutter_bird_image_server/server.js:28]: This only needs to be executed when you want to specify a port when starting locally, so currently `process.env.NODE_ENV === 'production'` is not necessary.
(When deployed to vercel, process.env.NODE_ENV will be production.)

**Comment by shiki-tak** [comment]
@Mdaiki0730 PTAL comments.

And the problem of opening the app store without starting metamask may be due to a problem with the device or metamask settings. Therefore, as a result of the meeting, we decided to deal with it separately from this PR.

File: Dockerfile (hunk 1 above)
Comment (Mdaiki0730) [reply Dockerfile]: The lock file will be generated even if it is not included, but if you include it, you can specify the version of the dependencies, which will increase the reproducibility of the environment.
This is a file to eliminate environmental differences, so how about include it?

> Describe a single representation of a dependency tree such that teammates, deployments, and continuous integration are guaranteed to install exactly the same dependencies.
https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json

Comment (shiki-tak) [reply Dockerfile]: Ok, I include it again, Thanks!

**Review by Mdaiki0730 — APPROVED** [review]

**Pull request merged**